REST API: Handle edge cases for Jetpack setup

itsmeichigo opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes #8919
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR fixes 2 edge cases for Jetpack setup discussed in [pe5sF9-1oh-p2]:
- When a user sets up WCPay for there non-Jetpack site, tapping on the install button on the Jetpack benefit modal will authenticate them with the connected WPCom account and show the install flow for JCP right after the sync is done. Please note: the install flow for JCP doesn't work currently due to the outdated code in JCP, so please ignore the issue until this is fixed on the backend. (Internal discussion: p1680512610989549-slack-CBG1CP4EN)
- When a site has a custom URL for its admin page, Jetpack connection will fail with error code 404. The fix is to detect the error and show the error screen with a suggestion to connect on a browser or contact support.

Also: the feature flag for Jetpack setup is enabled to release this feature for 13.1.

## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->
1. Testing JCP install
- Prerequisite: make sure that you have a self-hosted Woo store without Jetpack.
- Log in to the test store if needed.
- Install, activate, and connect WCP to your WPCom account.
- Navigate to the Jetpack setup flow by tapping the Jetpack benefit banner on the dashboard screen or go to Menu > Settings > Install Jetpack.
- Log in to your WPCom account as part of the Jetpack setup flow. After the login succeeds, a loading screen should appear, and then you should be authenticated with the account.
- Notice that the setup flow for JCP will then be presented.

2. Testing custom admin page.
- Make sure that WCPay is deactivated on your test store.
- Install and activate [WP Hide](https://wordpress.org/plugins/wp-hide-security-enhancer/) plugin on your store.
- On wp-admin left side bar navigate to WP Hide > Hide Admin. Update a custom URL and enable "Block default Admin Url"
<img width="1087" alt="Screenshot 2023-04-04 at 09 35 33" src="https://user-images.githubusercontent.com/5533851/229679826-54444b80-d040-4eef-a4cd-7d8fbf9dfe97.png">

- On the app, navigate to the Jetpack setup flow by tapping the Jetpack benefit banner on the dashboard screen or go to Menu > Settings > Install Jetpack.
- During the connection step of the setup flow, notice that the connection web view is dismissed and the setup screen should show an error about the connection.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->


https://user-images.githubusercontent.com/5533851/229794970-11eb2627-d95c-4720-8229-32bcb7cea40e.mp4





---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

itsmeichigo committed:
Show Jetpack install flow for JCP after synchronizing if necessary

itsmeichigo committed:
Add new method in AuthenticatedWebViewModel for deciding policy of navigation response

itsmeichigo committed:
Implement new protocol method for navigation response policy decision on AuthenticatedWebViewController

itsmeichigo committed:
Update JetpackConnectionWebViewModel to implement the new protocol method

itsmeichigo committed:
Add content for the connection error

itsmeichigo committed:
Show connection error when encountering 404 during connection

itsmeichigo committed:
Merge branch 'trunk' into feat/8919-edge-cases

itsmeichigo committed:
Enable feature flag for jetpack setup with application password

itsmeichigo committed:
Update release notes

itsmeichigo committed:
Fix typo for localized string comments

selanthiraiyan commented:
> 2. Testing custom admin page. 
> During the connection step of the setup flow, notice that the connection web view is dismissed and the setup screen should show an error about the connection.

Just to confirm, I see the following error. 

<img src="https://user-images.githubusercontent.com/524475/229701832-6824215e-6677-4f80-8a5c-9bacef305aa6.png" width="320"/>

selanthiraiyan commented on WooCommerce/Classes/Authentication/Navigation Exceptions/JetpackConnectionWebViewModel.swift:
```diff
@@ -10,30 +10,34 @@ final class JetpackConnectionWebViewModel: AuthenticatedWebViewModel {
          stores: StoresManager = ServiceLocator.stores,
          analytics: Analytics = ServiceLocator.analytics,
          completion: @escaping () -> Void,
+         onFailure: @escaping () -> Void = {},
```
Just a 💭

Should we try to clarify that `onFailure` is fired when we encounter 404? I am asking this because we specifically treat this as 404 [here](https://github.com/woocommerce/woocommerce-ios/pull/9354/files#diff-5b2d3d7f53b79702e24015df8d058f2c45f0b6ba5e79eba6db48da378af75cccR134-R139)

itsmeichigo replied:
Good point! I updated the logic to send the error code from the web view model and display it on the error screen as is in 5c2ba687415833a384790c0b650ce79e35b87e32.

selanthiraiyan commented on WooCommerce/Classes/Authentication/Navigation Exceptions/JetpackConnectionWebViewModel.swift:
```diff
@@ -57,6 +61,18 @@ final class JetpackConnectionWebViewModel: AuthenticatedWebViewModel {
+        }
+        await MainActor.run { [weak self] in
+            guard let self else { return }
+            self.isFailure = true
```
Super nit: 

My understanding is we introduced `isFailure` to allow dismissal of the view. 

Could we instead keep using `isCompleted` and set it as `true`? Maybe we can rename it into something like `isProcessing`. 

Please feel free to keep the new `isFailure` variable if you think it adds to readability.

itsmeichigo replied:
That makes sense, I unified `isCompleted` and `isFailure` into `shouldIgnoreDismissalHandling` for clarity in b12441481235efea7b46894c4f3b8368454ecef0.

selanthiraiyan reviewed (approved):
This works as expected. 🚀 

I left a couple of non-blocking comments.

itsmeichigo committed:
Send error code to display on the error screen

itsmeichigo committed:
Use shouldIgnoreDismissalHandling to skip tracking and calling dismissal handler

itsmeichigo commented:
@selanthiraiyan Regarding the error screen you saw above - that's the error screen for Jetpack setup for JCP sites. Could you test again by deactivating WCPay and ensuring that Jetpack is unavailable for the site?

Regarding the failure of Jetpack installation for JCP, the cause is probably on the backend, according to the Slack thread I linked to in the description.

selanthiraiyan commented:
Should I install and uninstall WCPay before testing the following case? 

> Could you test again by deactivating WCPay and ensuring that Jetpack is unavailable for the site?

I created a new JN site. Installed WP-Hide and followed the steps to customize the admin URL. 
<img width="1570" alt="Screenshot 2023-04-04 at 3 44 23 PM" src="https://user-images.githubusercontent.com/524475/229761218-97c2189f-b5bf-4f70-a108-f49a2796f33f.png">

I see the following error. I am trying to confirm because I don't see the error screen in [the video recording](https://user-images.githubusercontent.com/5533851/229680111-b8fc6f76-fa10-4cbc-9862-9488c606f46f.mp4) from the PR description. 
<img src="https://user-images.githubusercontent.com/524475/229761322-831ae89c-62f7-4827-8068-9e86ef079e2d.png" width="320"/>

itsmeichigo committed:
Fix issue checking for error code 404

itsmeichigo committed:
Add new tests for JetpackConnectionWebViewModel

itsmeichigo commented:
> Should I install and uninstall WCPay before testing the following case?

@selanthiraiyan WCPay is a separate edge case - from what I understand you are testing the second case which doesn't require WCPay to be installed.

> I see the following error. I am trying to confirm because I don't see the error screen in [the video recording](https://user-images.githubusercontent.com/5533851/229680111-b8fc6f76-fa10-4cbc-9862-9488c606f46f.mp4) from the PR description.

The screencast I posted needed to be corrected, I updated the correct one in the PR description - my bad.
The screen you see is correct, but the error code is not. I fixed that in 9d07768c47af067223d27b6e729385e86abd2f13 and added new tests in b9cc2c3a75792ee545a8b3820d7e2ffbae0cbdaa. This PR is ready for another look - thanks.

selanthiraiyan reviewed (approved):
Thank you for the updates! 🚀
